ceph/ceph.go

```diff
@@ -0,0 +1,253 @@
+	"fmt"
+	"os/exec"
+	"strconv"
+	"sync"
+
+	"github.com/Sirupsen/logrus"
+
+	. "github.com/rancher/convoy/convoydriver"
+	"github.com/rancher/convoy/util"
+	"github.com/rancher/convoy/util/fs"
+)
+
+var (
+	log = logrus.WithFields(logrus.Fields{"pkg": "ceph"})
+)
+
+const (
+	driverName = "ceph"
+
+	cephDefaultVolumeSize = "ceph.defaultvolumesize"
+	cephDefaultEncrypted  = "ceph.defaultencrypted"
+
+	defaultVolumeSize = "10G"
+	defaultEncrypted  = "false" // Currently unused, but may be supported in future
+	defaultFSType     = "ext4"
+)
+
+type Driver struct {
+	mutex   *sync.RWMutex
+	volumes map[string]*Volume
+	*Device
+}
+
+type Device struct {
+	Root              string
+	DefaultVolumeSize int64
+	DefaultEncrypted  bool
+}
+
+func (d *Driver) VolumeOps() (VolumeOperations, error) {
+	return d, nil
+}
+
+func (Driver) SnapshotOps() (SnapshotOperations, error) {
+	return nil, fmt.Errorf("Snapshot ops not supported")
+}
+
+func (Driver) BackupOps() (BackupOperations, error) {
+	return nil, fmt.Errorf("Backup ops not supported")
+}
+
+func (d *Driver) Info() (map[string]string, error) {
+	return map[string]string{
+		"name": d.Name(),
+	}, nil
+}
+
+func init() {
+	if err := Register(driverName, Init); err != nil {
+		panic(err)
+	}
+}
+
+func (*Driver) Name() string {
+	return "ceph"
+}
+
+// CreateVolume is empty as we create on mount
+func (d *Driver) CreateVolume(req Request) error {
+	return nil
+}
+
+// createVolume is used on mount to generate the volumes internal state
+func (d *Driver) createVolume(req Request) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	v, exists := d.volumes[req.Name]
+	if !exists {
+		v = &Volume{
+			Name:             req.Name,
+			Device:           "", // Will be set by Mount()
+			LUKSDevice:       "", // Will be set by Mount()
+			MountPointPrefix: d.Root,
+		}
+		d.volumes[req.Name] = v
+	}
+}
+
+func (d *Driver) DeleteVolume(req Request) error {
+	currentImageMap, err := d.getCurrentVolumes()
+	if err != nil {
+		return err
+	}
+	// if the image is still mapped then don't delete the volume from Convoy internal state
+	if _, exists := currentImageMap[req.Name]; exists {
+		return nil
+	}
+	if _, exists := d.volumes[req.Name]; exists {
+		delete(d.volumes, req.Name)
+	}
+	return nil
+}
+
+func (d *Driver) MountVolume(req Request) (string, error) {
+	// if the volume state does not exist then generate it
+	if _, exists := d.volumes[req.Name]; !exists {
+		d.createVolume(req)
+	}
+	volume := d.volumes[req.Name]
+	// Map the volume
+	var err error
+	if _, err := volume.Map(req.Name, d.DefaultVolumeSize); err != nil {
+		return "", err
+	}
+	defer func() {
+		if err != nil {
+			if err = volume.unmapCephVolume(); err != nil {
+				log.Debug(err)
+			}
+		}
+	}()
+
+	fsType, err := volume.detectFS()
+	if err == fs.ErrNoFilesystemDetected {
+		if err = volume.createFS(); err != nil {
+			return "", err
+		}
+		fsType = defaultFSType
+	} else if err != nil {
+		return "", err
+	}
+	if fsType == cryptoLuksFsType {
```

Comment:
That is ugly :-) Any way to use a driver option like "encrypted" instead of overloading the file system ? What if we want to allow multiple file systems ? Don't you want to also allow the creation of arbitrary FS ?

Reply:
Sadly we cannot pass options to the volume driver from docker run (can do that when running docker volume create tho).
I'm starting to test this and possibly revive it, maintaining the patches for docker is becoming a pain.